lib/executor/ramping_vus.go

```diff
@@ -493,154 +493,194 @@ var _ lib.Executor = &RampingVUs{}
-		StartTime:  startTime,
-		ProgressFn: progressFn,
-	})
-
-	vuHandles := make([]*vuHandle, maxVUs)
-	for i := uint64(0); i < maxVUs; i++ {
-		vuHandle := newStoppedVUHandle(
-			maxDurationCtx, getVU, returnVU, vlv.nextIterationCounters,
-			&vlv.config.BaseConfig, vlv.logger.WithField("vuNum", i))
-		go vuHandle.runLoopsIfPossible(runIteration)
-		vuHandles[i] = vuHandle
+	for i := uint64(0); i < rs.startMaxVUs; i++ {
+		rs.vuHandles[i] = newStoppedVUHandle(
+			ctx, getVU, returnVU, rs.executor.nextIterationCounters,
+			&rs.executor.config.BaseConfig, rs.executor.logger.WithField("vuNum", i))
 	}
+}
 
-	// 0 <= currentScheduledVUs <= currentMaxAllowedVUs <= maxVUs
-	var currentScheduledVUs, currentMaxAllowedVUs uint64
-
-	handleNewScheduledVUs := func(newScheduledVUs uint64) {
-		if newScheduledVUs > currentScheduledVUs {
-			for vuNum := currentScheduledVUs; vuNum < newScheduledVUs; vuNum++ {
-				_ = vuHandles[vuNum].start() // TODO handle error
-			}
+func (rs rampingVUsRunState) handleVUs(ctx context.Context) {
+	// iterate over stepsRaw and stepsGraceful in order by TimeOffset
+	// giving stepsRaw precedence.
+	// we stop iterating once stepsRaw are over as we need to run the remaining
+	// stepsGraceful concurrently while waiting for VUs to stop in order to not wait until
+	// the end of gracefulStop (= maxDur-regDur) timeouts
+	var (
+		handleNewMaxAllowedVUs = rs.maxAllowedVUsHandlerStrategy()
+		handleNewScheduledVUs  = rs.scheduledVUsHandlerStrategy()
+		wait                   = waiter(ctx, rs.started)
+	)
+	for len(rs.stepsRaw) > 0 {
+		var (
+			r, g   = rs.stepsRaw[0], rs.stepsGraceful[0]
+			gfirst = r.TimeOffset > g.TimeOffset
+			offset = g.TimeOffset
+		)
+		if !gfirst {
+			offset = r.TimeOffset
+		}
+		if wait(offset) {
+			return
+		}
+		if gfirst {
+			handleNewMaxAllowedVUs(g)
+			rs.stepsGraceful = rs.stepsGraceful[1:]
 		} else {
-			for vuNum := newScheduledVUs; vuNum < currentScheduledVUs; vuNum++ {
-				vuHandles[vuNum].gracefulStop()
-			}
+			handleNewScheduledVUs(r)
+			rs.stepsRaw = rs.stepsRaw[1:]
+		}
+	}
+}
+
+// TODO: removing this has no effect on tests?
+func (rs rampingVUsRunState) handleRemainingVUs(ctx context.Context) {
+	var (
+		handleNewMaxAllowedVUs = rs.maxAllowedVUsHandlerStrategy()
+		wait                   = waiter(ctx, rs.started)
+	)
+	for _, s := range rs.stepsGraceful {
+		if wait(s.TimeOffset) {
+			return
 		}
-		currentScheduledVUs = newScheduledVUs
+		handleNewMaxAllowedVUs(s)
 	}
+}
 
-	handleNewMaxAllowedVUs := func(newMaxAllowedVUs uint64) {
-		if newMaxAllowedVUs < currentMaxAllowedVUs {
-			for vuNum := newMaxAllowedVUs; vuNum < currentMaxAllowedVUs; vuNum++ {
-				vuHandles[vuNum].hardStop()
+func (rs rampingVUsRunState) maxAllowedVUsHandlerStrategy() func(lib.ExecutionStep) {
+	var max uint64
+	return func(graceful lib.ExecutionStep) {
+		cur := graceful.PlannedVUs
+		if cur < max {
+			for n := cur; n < max; n++ {
+				rs.vuHandles[n].hardStop()
 			}
 		}
```

Comment:
Agreed with the algorithm change. But for the renaming part, I instead shy away from using reserved words like `new`. What is the reason for this renaming?

Reply:
IDK yet, is this because of the algorithm change but `make check` gave me errors after I change the algorithm. I also tried without `max - 1` with no luck... Hmm.

```
--- FAIL: TestEmittedMetricsWhenScalingDown (13.93s)
    engine_test.go:910: 
                Error Trace:    engine_test.go:910
                Error:          Not equal: 
                                expected: 3
                                actual  : 4
                Test:           TestEmittedMetricsWhenScalingDown
    engine_test.go:914: 
                Error Trace:    engine_test.go:914
                Error:          Not equal: 
                                expected: 8
                                actual  : 9
                Test:           TestEmittedMetricsWhenScalingDown
    engine_test.go:925: The data_received sum should be in the interval [120000.000000, 124000.000000] but was 136881.000000
    engine_test.go:934: 
                Error Trace:    engine_test.go:934
                Error:          Not equal: 
                                expected: 3
                                actual  : 4
                Test:           TestEmittedMetricsWhenScalingDown
```

Reply:
But the following one works (as expected):

```golang
cur := graceful.PlannedVUs
for n := cur; n < max; n++ {
	rs.vuHandles[n].hardStop()
}
max = cur
```

Reply:
Btw, the algorithm change causes the following errors.

Are VUs can be stopped in reverse order?

```
--- FAIL: TestConstantArrivalRateRunCorrectTiming (0.00s)
    --- FAIL: TestConstantArrivalRateRunCorrectTiming/segment_1/3:2/3_sequence_<nil> (2.00s)
        constant_arrival_rate_test.go:222: 
                Error Trace:    constant_arrival_rate_test.go:222
                                                        common_test.go:40
                                                        minirunner.go:230
                                                        helpers.go:88
                                                        ramping_arrival_rate.go:549
                                                        asm_amd64.s:1581
                Error:          Max difference between 2021-09-29 19:14:25.835391 +0300 +03 m=+0.316877023 and 2021-09-29 19:14:25.856556 +0300 +03 m=+0.338052136 allowed is 12ms, but difference was -21.175113ms
                Test:           TestConstantArrivalRateRunCorrectTiming/segment_1/3:2/3_sequence_<nil>
                Messages:       5 expectedTime 260ms
    --- FAIL: TestConstantArrivalRateRunCorrectTiming/segment_1/3:2/3_sequence_0,1/3,2/3,1 (2.00s)
        constant_arrival_rate_test.go:222: 
                Error Trace:    constant_arrival_rate_test.go:222
                                                        common_test.go:40
                                                        minirunner.go:230
                                                        helpers.go:88
                                                        ramping_arrival_rate.go:549
                                                        asm_amd64.s:1581
                Error:          Max difference between 2021-09-29 19:14:25.834746 +0300 +03 m=+0.316232200 and 2021-09-29 19:14:25.856528 +0300 +03 m=+0.338024187 allowed is 12ms, but difference was -21.791987ms
                Test:           TestConstantArrivalRateRunCorrectTiming/segment_1/3:2/3_sequence_0,1/3,2/3,1
                Messages:       5 expectedTime 260ms
    --- FAIL: TestConstantArrivalRateRunCorrectTiming/segment_1/6:1/2_sequence_<nil> (2.00s)
        constant_arrival_rate_test.go:222: 
                Error Trace:    constant_arrival_rate_test.go:222
                                                        common_test.go:40
                                                        minirunner.go:230
                                                        helpers.go:88
                                                        ramping_arrival_rate.go:549
                                                        asm_amd64.s:1581
                Error:          Max difference between 2021-09-29 19:14:25.835546 +0300 +03 m=+0.317032505 and 2021-09-29 19:14:25.856503 +0300 +03 m=+0.337999310 allowed is 12ms, but difference was -20.966805ms
                Test:           TestConstantArrivalRateRunCorrectTiming/segment_1/6:1/2_sequence_<nil>
                Messages:       5 expectedTime 260ms
    --- FAIL: TestConstantArrivalRateRunCorrectTiming/segment_1/6:1/2_sequence_1/6,1/2 (2.00s)
        constant_arrival_rate_test.go:222: 
                Error Trace:    constant_arrival_rate_test.go:222
                                                        common_test.go:40
                                                        minirunner.go:230
                                                        helpers.go:88
                                                        ramping_arrival_rate.go:549
                                                        asm_amd64.s:1581
                Error:          Max difference between 2021-09-29 19:14:25.834704 +0300 +03 m=+0.316189855 and 2021-09-29 19:14:25.856586 +0300 +03 m=+0.338081978 allowed is 12ms, but difference was -21.892123ms
                Test:           TestConstantArrivalRateRunCorrectTiming/segment_1/6:1/2_sequence_1/6,1/2
                Messages:       5 expectedTime 260ms
FAIL
FAIL    go.k6.io/k6/lib/executor        34.960s
```

Reply:
> I instead shy away from using reserved words like new. What is the reason for this renaming?

ah ... yes it's reserved ... than something that has the same meaning as `new`. And the reason is that the action we are doing is the one thing that changes the actual thing we care about so the argument is the "new" value, while what we have is what is the "current" value.

Reply:
This looks as some flaky tests and the change worked for me locally, so maybe you are hitting the flaky tests more :thinking: . Maybe rerun them a bit on master (with `-count 1 ./... -short` to not use the cache, run everything so it's flaky, and not run the tc39 tests (short) as you don't need them ;) ) and open an issue if it keeps happening with some details about your system

Reply:
Thx for the tip. That's why I've been running the tests with the `-v` flag to prevent `go test` caching them. Btw, it's not always on my system, the CI flaky-fails as well.

Reply:
AFAIK `-v` doesn't prevent the usage of cache and using it locally it definitely doesn't do for my setup